client/allocrunner/taskrunner/task_runner.go

```diff
@@ -444,17 +479,76 @@ func (tr *TaskRunner) runDriver() error {
 	//TODO mounts and devices
 	//XXX Evaluate and encode driver config
 
-	// Start the job
-	handle, net, err := tr.driver.StartTask(taskConfig)
+	var handle *drivers.TaskHandle
+	var net *cstructs.DriverNetwork
+	var err error
+
+	// Check to see if a task handle was restored
+	tr.localStateLock.RLock()
+	handle = tr.localState.TaskHandle
+	net = tr.localState.DriverNetwork
+	tr.localStateLock.RUnlock()
+
+	if handle != nil {
+		tr.logger.Trace("restored handle; recovering task", "task_id", handle.Config.ID)
+		if err := tr.driver.RecoverTask(handle); err != nil {
+			tr.logger.Error("error recovering task; destroying and restarting",
+				"error", err, "task_id", handle.Config.ID)
+
+			// Clear invalid task state
+			tr.localStateLock.Lock()
+			tr.localState.TaskHandle = nil
+			tr.localState.DriverNetwork = nil
+			tr.localStateLock.Unlock()
+
+			// Try to cleanup any existing task state in the plugin before restarting
+			if err := tr.driver.DestroyTask(handle.Config.ID, true); err != nil {
+				// Ignore ErrTaskNotFound errors as ideally
+				// this task has already been stopped and
+				// therefore doesn't exist.
+				if err != drivers.ErrTaskNotFound {
+					tr.logger.Warn("error destroying unrecoverable task",
+						"error", err, "task_id", handle.Config.ID)
+				}
+
+			}
+
+			goto START
+		}
+
+		// Update driver handle on task runner
+		tr.setDriverHandle(NewDriverHandle(tr.driver, handle.Config.ID, tr.Task(), net))
+
+		// Ensure running state is persisted but do *not* append a new
+		// task event as restoring is a client event and not relevant
+		// to a task's lifecycle.
+		if err := tr.updateStateImpl(structs.TaskStateRunning); err != nil {
+			tr.logger.Warn("error persisting task state", "error", err)
+		}
+		return nil
+	}
+
+START:
+	// Start the job if there's no existing handle (or if RecoverTask failed)
+	handle, net, err = tr.driver.StartTask(taskConfig)
 	if err != nil {
 		return fmt.Errorf("driver start failed: %v", err)
 	}
 
 	tr.localStateLock.Lock()
 	tr.localState.TaskHandle = handle
+	tr.localState.DriverNetwork = net
+	if err := tr.stateDB.PutTaskRunnerLocalState(tr.allocID, tr.taskName, tr.localState); err != nil {
+		//TODO Nomad will be unable to restore this task; try to kill
+		//     it now and fail? In general we prefer to leave running
+		//     tasks running even if the agent encounters an error.
+		tr.logger.Warn("error persisting local task state; may be unable to restore after a Nomad restart",
```

Comment:
I think anywhere that has an error persisting state needs to kick a singleton, background persist state routine that tries to save all local state on a retry loop. Right now its edge triggered on state transition which means if it fails to save on the last trigger, nothing tries to store the state